### PR TITLE
fix: multiple fixes to the agent-rs, most notably having 5 min limit on the retry loop while polling 

### PR DESCRIPTION
**Pull request by rumenov**
# Description

We want to use the public agent-rs as much as possible in the core protocol code and remove the existing custom solutions we have. For this purpose we may need to bridge some differences. For example there is no point in polling for a response more than 5 minutes since the replica will GC any ingress messages within 5 min anyway.

The MR also updates versions of some dependencies as well.

Fixes # (issue)

# How Has This Been Tested?

CI

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

### CODE REVIEW

File: ic-agent/src/agent/agent_error.rs [hunk 1]
```diff
@@ -127,9 +127,9 @@ pub enum AgentError {
     #[error("The status response did not contain a root key.  Status: {0}")]
     NoRootKeyInStatus(Status),
 
-    /// Could not read the replica root key.
-    #[error("Could not read the root key")]
-    CouldNotReadRootKey(),
+    /// An operation failed due to a poisoned type.
+    #[error("Could not perform the requested operation due to a poisoned type")]
+    Poisoned(),
```
Comment (adamspofford-dfinity) [inline ic-agent/src/agent/agent_error.rs:132]: A poison state shouldn't be surfaced as an error at all. A call in the stack that does not have the ability to reset the mutex should receive it as a panic; it exists to propagate panics from one thread to another and if we do not provide the ability to recover from the other panic, the result should be unwrapped. Either way there should be no such error variant.

File: ic-agent/src/agent/mod.rs [hunk 2]
```diff
@@ -263,19 +268,23 @@ impl Agent {
...
                 .unwrap_or_else(|| Duration::from_secs(300)),
-            root_key: Arc::new(RwLock::new(Some(IC_ROOT_KEY.to_vec()))),
+            root_key: Arc::new(RwLock::new(IC_ROOT_KEY.to_vec())),
             transport: config
                 .transport
                 .ok_or_else(AgentError::MissingReplicaTransport)?,
+            backoff_initial_interval: Duration::from_millis(500),
+            backoff_max_elapsed_time: Duration::from_secs(60 * 2),
+            backoff_max_interval: Duration::from_secs(1),
+            backoff_multiplier: 1.4,
```
Comment (adamspofford-dfinity) [inline ic-agent/src/agent/mod.rs:278]: These attributes should also be surfaced in AgentBuilder.

File: ic-agent/src/agent/agent_error.rs (hunk 1 above)
Comment (rumenov) [reply ic-agent/src/agent/agent_error.rs]: Sure, this is even better. I just decided to keep the existing semantics but I do agree if we unwrap the lock it is better.

**Comment by rumenov** [comment]
Thanks for the review @adamspofford-dfinity , can you please take another look

File: ic-agent/src/agent/builder.rs [hunk 3]
```diff
@@ -83,12 +84,27 @@ impl AgentBuilder {
...
-                ..self.config
-            },
-        }
+    pub fn with_ingress_expiry(mut self, ingress_expiry: Option<std::time::Duration>) -> Self {
+        self.config.ingress_expiry = ingress_expiry;
+        self
+    }
+
+    /// Sets the initial backoff interval used for polling the IC when awaiting a response.
+    pub fn set_backoff_initial_interval(mut self, initial_interval: Duration) -> Self {
```
Comment (adamspofford-dfinity) [inline ic-agent/src/agent/builder.rs:93]: These should be named `with_`, not `set_`.

File: ic-agent/src/agent/mod.rs [hunk 4]
```diff
@@ -260,33 +265,20 @@ impl Agent {
...
-    /// NOTE: if you change the identity while having update calls in
-    /// flight, you will not be able to [Agent::poll] the status of these
-    /// messages.
-    pub fn set_identity<I>(&mut self, identity: I)
-    where
-        I: 'static + Identity,
-    {
-        self.identity = Arc::new(identity);
-    }
-
```
Comment (adamspofford-dfinity) [inline ic-agent/src/agent/mod.rs:289]: unintentional deletion? set_identity is here on purpose.

Comment (rumenov) [reply ic-agent/src/agent/mod.rs:289]: Actually it was intentional. Do we really need to change the identity and transport of an already existing agent ? Especially given the working message in the comments ?

Comment (adamspofford-dfinity) [reply ic-agent/src/agent/mod.rs:289]: Yes. For the identity, a developer does not necessarily want to use the same identity for every call (and `poll` is little-used); for the transport, it is used in `dfx canister sign` specifically to wrap an existing agent.

Comment (rumenov) [reply ic-agent/src/agent/mod.rs]: ok, I reverted it for now. I can imagine that a user would like to set a new identity for different requests. However, setting a different transport may be a bit off. I would imagine if someone needs a new transport they would just create a new agent.

**Review by adamspofford-dfinity — APPROVED** [review]

**Pull request merged**